File: forms/dyslexia-assessment/back-end-with-loco/src/engine/dyslexia_grader.rs
use super::dyslexia_rules::all_rules;
use super::flagged_issues::detect_additional_flags;
use super::types::{AdditionalFlag, AssessmentData, DomainScore, GradingResult};
use super::utils::{max_severity, score_severity};

/// Pure function: grades a Dyslexia Assessment battery.
///
/// Returns a [`GradingResult`] containing:
/// - overall severity (driven by the *lowest* non-null standardised score —
///   i.e. the most-impaired domain — matching the JS reference engine)
/// - one `DomainScore` per rule
/// - additional flags (clinician-facing alerts)
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let additional_flags = detect_additional_flags(data);
    grade_with_flags(data, additional_flags, timestamp)
}

/// Internal helper that lets callers supply pre-built flags & a deterministic
/// timestamp (used in tests).
pub fn grade_with_flags(
    data: &AssessmentData,
    additional_flags: Vec<AdditionalFlag>,
    timestamp: String,
) -> GradingResult {
    let mut domain_scores: Vec<DomainScore> = Vec::new();
    let mut overall: String = "none".to_string();
    let mut lowest_score: Option<i32> = None;
    let mut answered_count: u32 = 0;

    for rule in all_rules() {
        let score = (rule.evaluate)(data);
        let severity = score_severity(score);
        domain_scores.push(DomainScore {
            id: rule.id.to_string(),
            category: rule.category.to_string(),
            description: rule.description.to_string(),
            score,
            severity: severity.clone(),
        });
        if let Some(s) = score {
            answered_count += 1;
            overall = max_severity(&overall, &severity);
            lowest_score = match lowest_score {
                None => Some(s),
                Some(cur) if s < cur => Some(s),
                Some(cur) => Some(cur),
            };
        }
    }

    GradingResult {
        overall_severity: overall,
        lowest_score,
        answered_count,
        domain_scores,
        additional_flags,
        timestamp,
    }
}

File: forms/dyslexia-assessment/front-end-with-svelte/src/lib/engine/dyslexia-grader.ts
import type { AssessmentData, DomainScore, Severity, GradingResult } from './types';
import { dyslexiaRules } from './dyslexia-rules';
import { scoreSeverity, maxSeverity } from './utils';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Pure function: evaluates the dyslexia battery against the supplied data.
 *
 * The overall severity is driven by the lowest non-null standardised score
 * (the most-impaired domain), matching the per-domain rules. If no domain has
 * been answered, the overall severity is 'none' and `lowestScore` is null.
 */
export function gradeDyslexia(data: AssessmentData): GradingResult {
	const domainScores: DomainScore[] = [];
	let overall: Severity = 'none';
	let lowestScore: number | null = null;
	let answeredCount = 0;

	for (const rule of dyslexiaRules) {
		let score: number | null = null;
		try {
			score = rule.evaluate(data);
		} catch (e) {
			console.warn(`Dyslexia rule ${rule.id} evaluation failed:`, e);
		}
		const severity = scoreSeverity(score);
		domainScores.push({
			id: rule.id,
			category: rule.category,
			description: rule.description,
			score: score === undefined ? null : score,
			severity
		});
		if (score !== null && score !== undefined && !Number.isNaN(score)) {
			answeredCount++;
			overall = maxSeverity(overall, severity);
			if (lowestScore === null || score < lowestScore) {
				lowestScore = score;
			}
		}
	}

	return {
		overallSeverity: overall,
		lowestScore,
		answeredCount,
		domainScores,
		additionalFlags: detectAdditionalFlags(data),
		timestamp: new Date().toISOString()
	};
}
